Fix PathFollow tests, Add PathFollow3D forward test

raulsntos opened the pull request:
- Fixes the tests for `PathFollow2D` and `PathFollow3D` so they compile and add them to `test_main.cpp` because they weren't being used before.
- Adds a test for the the forward direction of `PathFollow3D` to prevent regressions like #51342 in the future.

---

**NOTE:** I've marked this PR ready for review since the CI passes but I'd like to know what to do with the _remove point test_ and if the custom `is_equal_approx` is acceptable, then I'll squash the commits.

raulsntos committed:
Fix PathFollow tests, Add forward vector test to PathFollow3D

raulsntos commented on tests/test_path_follow_3d.h:
```diff
@@ -39,25 +39,26 @@
-	const PathFollow3D *path_follow_3d = memnew(PathFollow3D);
+	PathFollow3D *path_follow_3d = memnew(PathFollow3D);
 	path->add_child(path_follow_3d);
+	path_follow_3d->notification(Node::NOTIFICATION_ENTER_TREE);
```
`PathFollow3D` sets its `path` property to its parent `Path3D` on entering the tree but the `path` parent is not in the tree so the notification `NOTIFICATION_ENTER_TREE` is never emitted, I'm calling this method manually on every test. Is this appropriate for a test or too hacky?

raulsntos commented on tests/test_path_follow_2d.h:
```diff
@@ -81,15 +81,15 @@ TEST_CASE("[PathFollow2D] Sampling with unit offset") {
 	curve->add_point(Vector2(0, 100));
 	curve->add_point(Vector2(0, 0));
-	const Path2D *path = memnew(Path2D);
+	Path2D *path = memnew(Path2D);
```
The `PathFollow2D` tests crash on instantiating `Path2D` since it's a `CanvasItem` and that constructor calls `RenderingServer::get_singleton()->canvas_item_create()` but that singleton is null since it seems the `RenderingServer` is never initialized for tests. Not sure what to do here.

Paulb23 commented:
For the `SceneTree` and `RenderingSever`, take a look at #50375.

raulsntos commented:
The tests are failing because I get `Vector3(99.93983, 0.050718, 0)` instead of `Vector3(100, 0, 0)` probably because of float precision or the curve bake quality but I tried to reduce the `bake_interval` and even though it gets closer to `Vector3(100, 0, 0)` it's still not enough and if I reduce it too much the tests crash probably because it runs out of memory.

akien-mga commented:
You can use `approx_equal` if it's close enough, or compare manually to a big enough epsilon that takes precision into account.

raulsntos commented:
I'm using `Vector3::is_equal_approx` which, unlike `Math::is_equal_approx`, does not have an overload with a `tolerance` parameter and therefore uses `CMP_EPSILON`.

The difference between `Vector3(99.93983, 0.050718, 0)` and `Vector3(100, 0, 0)` is not huge but it's not tiny either, so I'm not sure if modifying the tests' precision to pass would be acceptable. Maybe the difference is caused by an underlying issue and we'd be masking it, no?

akien-mga commented:
> The difference between `Vector3(99.93983, 0.050718, 0)` and `Vector3(100, 0, 0)` is not huge but it's not tiny either, so I'm not sure if modifying the tests' precision to pass would be acceptable. Maybe the difference is caused by an underlying issue and we'd be masking it, no?

Yes, that should be investigated.

raulsntos commented:
I've extracted the Curve fixes to PR #58152 so they can be merged independently.

reduz commented:
Whats the status of this?

raulsntos commented:
It's done but not sure what to do about the last 2 commits:

- Is the custom `is_equal_approx` method acceptable? Otherwise, with the default tolerance, the tests don't pass.
- What is the expected behavior of removing a point? I commented the test because that's not how it currently works. I can remove the test, change the test or change the behavior.

When those questions are resolved I have to squash the commits and rebase.

YuriSizov commented:
This probably needs at least some changes after the recent PathFollow changes regarding forward direction.

raulsntos commented:
- While rebasing I found a bug in `PathFollow3D`, I opened https://github.com/godotengine/godot/pull/78280 to fix it. I will rebase this PR after the other one is merged.
- It seems the tests regarding forward direction fail unless I increase the tolerance even more, but at this point I feel like this might be a bug in `PathFollow3D` or `Curve3D`.

Geometror commented:
Could you do a quick rebase?

AThousandShips commented on tests/test_main.cpp:
```diff
@@ -90,6 +90,8 @@
 #include "tests/scene/test_gradient.h"
 #include "tests/scene/test_path_3d.h"
+#include "tests/scene/test_path_follow_2d.h"
+#include "tests/scene/test_path_follow_3d.h"
```
Move down to the `#ifndef _3D_DISABLED` part

AThousandShips commented on tests/scene/test_path_follow_2d.h:
```diff
@@ -32,205 +32,222 @@
-	path->set_curve(curve);
-	const PathFollow2D *path_follow_2d = memnew(PathFollow2D);
-	path->add_child(path_follow_2d);
+// TEST_CASE("[SceneTree][PathFollow2D] Removal of a point in curve") {
```
```suggestion
//TEST_CASE("[SceneTree][PathFollow2D] Removal of a point in curve") {
```
No space before commented out code, for all

raulsntos replied:
I've commented these tests because of https://github.com/godotengine/godot/pull/51372#discussion_r724466895. I don't think the behavior these tests intended to test was ever implemented in `PathFollow{2D,3D}` so we may want to remove these tests instead.

Geometror replied:
With the following adjustments it passes:p. We need to do the same immediate update as for the other test cases, and after the curve length has changed, the absolute progress value is outdated (and doesn't represent the ratio of 0.5 anymore).
```cpp
	path_follow_3d->set_progress_ratio(0.5);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(100, 0, 0), path_follow_3d->get_transform().get_origin()));

	curve->remove_point(1);
	path_follow_3d->set_progress_ratio(0.5);
	path_follow_3d->update_transform(true);
	CHECK_MESSAGE(
			is_equal_approx(Vector3(50, 50, 0), path_follow_3d->get_transform().get_origin()),
			"Path follow's position should be updated after removing a point from the curve");
```

raulsntos commented:
Rebased.

Also added `path_follow_3d->update_transform(true)` because https://github.com/godotengine/godot/pull/80233 made the transform update deferred.

It seems https://github.com/godotengine/godot/pull/80233 also broke the forward vector test and I'm not sure if the new behavior is correct. I also get a bunch of `ERROR: The target vector and up vector can't be parallel to each other.`

Geometror commented on tests/scene/test_path_follow_3d.h:
```diff
@@ -32,188 +32,284 @@
+	CHECK(is_equal_approx(Vector3(1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(50);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(100);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, 1, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(150);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, 1, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(200);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, 0, 1), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(250);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, 0, 1), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(300);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, -1, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(350);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, -1, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(400);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, -1, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	path_follow_3d->set_progress(450);
+	path_follow_3d->update_transform(true);
+	CHECK(is_equal_approx(Vector3(0, -1, 0), path_follow_3d->get_transform().get_basis().get_column(2)));
+
+	memdelete(path);
+}
```
I rebuilt the unit test in the editor and tested it with 4.1.3, 4.2.2 and 4.3 - exact same behavior, which is actually expected for the oriented rotation mode.
However, I think we should consider renaming the items of the RotationMenu enum since they are a bit confusing in my opinion (at least `ROTATION_ORIENTED`), but that's for another PR.
I'd adjust the tests the following way to avoid target vectors becoming parallel to the up vector. The checks were verified using a test project:

![grafik](https://github.com/user-attachments/assets/c4afa94b-c61c-4faa-85ae-d4f8887e1b7b)


```suggestion
TEST_CASE("[SceneTree][PathFollow3D] Calculate forward vector") {
	const real_t dist_cube_100 = 100 * Math::sqrt(3.0);
	Ref<Curve3D> curve = memnew(Curve3D);
	curve->add_point(Vector3(0, 0, 0));
	curve->add_point(Vector3(100, 0, 0));
	curve->add_point(Vector3(200, 100, -100));
	curve->add_point(Vector3(200, 100, 200));
	curve->add_point(Vector3(100, 0, 100));
	curve->add_point(Vector3(0, 0, 100));
	Path3D *path = memnew(Path3D);
	path->set_curve(curve);
	PathFollow3D *path_follow_3d = memnew(PathFollow3D);
	path->add_child(path_follow_3d);
	SceneTree::get_singleton()->get_root()->add_child(path);

	path_follow_3d->set_loop(false);
	path_follow_3d->set_rotation_mode(PathFollow3D::RotationMode::ROTATION_ORIENTED);

	path_follow_3d->set_progress(-50);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(-1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(0);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(-1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(50);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(-1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(100);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(-1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(100 + dist_cube_100 / 2);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(-0.577348, -0.577348, 0.577348), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(100 + dist_cube_100 - 0.01);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(-0.577348, -0.577348, 0.577348), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(250 + dist_cube_100);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(0, 0, -1), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(400 + dist_cube_100 - 0.01);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(0, 0, -1), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(400 + 1.5 * dist_cube_100);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(0.577348, 0.577348, 0.577348), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(400 + 2 * dist_cube_100 - 0.01);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(0.577348, 0.577348, 0.577348), path_follow_3d->get_transform().get_basis().get_column(2)));

	path_follow_3d->set_progress(500 + 2 * dist_cube_100);
	path_follow_3d->update_transform(true);
	CHECK(is_equal_approx(Vector3(1, 0, 0), path_follow_3d->get_transform().get_basis().get_column(2)));

	memdelete(path);
}
```

raulsntos replied:
If you say these are the expected values I'll take your word for it.

Also, even though it's a bit off-topic for this PR, you mentioned renaming the enum members but that would break compatibility. We can however improve the documentation and the property hint that determines how they are displayed in the inspector.

Geometror reviewed (approved):
Looks good to me!
We could refactor the checks to use `doctest::Approx`, but I'm working on a better way to integrate Godot's types with doctest, so I guess it's fine to merge this now.

akien-mga commented:
Seems good to me too. Let's squash and merge, and fine tune things in follow up PRs.

akien-mga commented:
Thanks!
